DOC-4291 Rewards administration guide (#101)
exoplatform/exo-documentation: docs/Installation.rst

.. _CustomizingEnvironmentVariables:

.. _RewardsInstall:

=============================
Rewards packages installation
=============================

.. note:: Only the administrator has the right to install and uninstall this application.

> The Rewards add-ons is supported for both Enterprise and Community editions.

Reviewer: ```suggestion
The Rewards add-ons are compatible with both Enterprise and Community editions.
```

Author: Done

> The Rewards add-ons are by default installed in both editions.

Reviewer: ```suggestion
Starting with version 5.3, Rewards add-ons are pre-installed by default in both editions.
```

Author: Done

> You can uninstall it using this command:

Reviewer: ```suggestion
You can uninstall them using these commands :
```

::

    ./addon uninstall exo-wallet
    ./addon uninstall exo-kudos
    ./addon uninstall exo-perk-store

> If you want to re-install it again, you should just run this command:

Reviewer: ```suggestion
If you want to re-install them again, you should just run these commands:
```

Author: Done

::

    ./addon install exo-wallet
    ./addon install exo-kudos
    ./addon install exo-perk-store

eXo Wallet addon
~~~~~~~~~~~~~~~~

> The wallet addon is used to add an ethereum blockchain wallet for each user.

Reviewer: ```suggestion
The Wallet addon is used to add an Ethereum blockchain token wallet for each user.
```

Author: Done

**Blockchain network settings**

> To install this addon, it's required to give access to eXo Platform Server the access permissions to communicate

Reviewer: ```suggestion
To setup this addon properly, it's required to give the eXo Platform Server the permissions to communicate
```

Author: Done

> with blockchain using websocket.

Reviewer: ```suggestion
with thus blockchain network using websocket.
```

Author: Deleted remark

The addon is configured by default to communicate with Ethereum Mainnet throught Infura using websocket protocol:

::

    exo.wallet.blockchain.networkId=1
    exo.wallet.blockchain.network.websocket=wss://mainnet.infura.io/ws/v3/a1ac85aea9ce4be88e9e87dad7c01d40
    exo.wallet.blockchain.network.http=https://mainnet.infura.io/v3/a1ac85aea9ce4be88e9e87dad7c01d40

The Websocket endpoint is used to:

>    -  Scan new mined blocks on blockchain to verify if there is a new mined transaction that involves a wallet of a recognized user.

Reviewer: ```suggestion
   -  Scan new blocks mined on the blockchain and to verify if there is any newly mined transaction that involves a wallet of a known user.
```

Author: Done

>       The periodicity can be configured using the following property (Default value: every hour):

Reviewer: ```suggestion
      The periodicity of the scan can be configured using the following property (Default value: every hour):
```

Author: Done

      ::

          exo.wallet.ContractTransactionVerifierJob.expression=0 0 * ? * * *

>    -  Checks all pending transactions sent using wallet application to update its state internally in internal store and to send notifications for receiver and sender.

Reviewer: ```suggestion
   -  Checks all pending transactions sent using the eXo Wallet application to update its internal state and to send notifications for receiver and sender.
```

Author: Done

      The check is made periodically and can be configured using the following property (Default value: every 10 seconds):

      ::

          exo.wallet.PendingTransactionVerifierJob.expression=0/10 * * * * ?

>    -  Send transactions on blockchain for a special wallet that is used to administrate all other wallets.

Reviewer: ```suggestion
   -  Transactions sent on the blockchain to a specific wallet address that is used to administrate all other wallets.
```

Author: Done

>       In fact this wallet is used to initialize other wallets with initial funds and to send reward amounts.

Reviewer: ```suggestion
      In fact this blockchain account is used to initialize user wallets with initial funds and to issue reward transactions.
```

Author: Done

> The HTTP endpoint is used to communicate with blockchain on Wallet UI.

Reviewer: ```suggestion
The HTTP endpoint is used to communicate with the blockchain in the eXo Wallet app.
```

Author: Done

> In fact, the users communicates directly with the blockchain to send transactions and don't use eXo Platform Server when sending transactions, unlike the * Admin wallet *.

Reviewer: ```suggestion
 Unlike the *Admin wallet*, the users' browser communicates directly with the blockchain to send transactions and it doesn't rely on the eXo Platform Server for issuing transactions.
```

Author: Done

> // TODO ajouter un schéma avec la blockchain, la UI d'eXo Wallet et le server eXo derrière qui communiquent ensemble (A DEMANDER AU DESIGNER de t'aider !!!!)

Reviewer: ```suggestion
// TODO add diagram of blockchain, with eXo Wallet UI and eXo server behind  that communicate together (NEED DESIGNER'S HELP)
```

Author: Done

**Admin wallet**

> The * Admin wallet * is created automatically on first eXo Platform server startup.

Reviewer: ```suggestion
The * Admin wallet * is created automatically during the first startup of the eXo Platform server.
```

Author: Done

> In order to be able to send transactions on blockchain, the private key of the * Admin wallet * is stored in internal database.

Reviewer: ```suggestion
In order to be able to send transactions on blockchain, the private key of the * Admin wallet * is stored in the internal database.
```

Author: Done

The private key is encrypted using two passwords:

     -  One password that is coming from a keystore file (see :ref:`Updating password encryption key <Security.UpdatePasswordEncryptionKey>`))
     -  And another password that can be configured in properties:

      ::

          exo.wallet.admin.key=changeThisKey

So, the * Admin wallet * can be used by multiple administrators of the platform to initialize other wallets.

> The private key of the admin wallet cannot be accessed by any user to avoid sharing its funds with a person.

Reviewer: ```suggestion
The private key of the admin wallet cannot be accessed by any user preventing to expose the funds outside the group of authorized administrators.
```

Author: Done

> To be an administrator of the platform that can manage wallets, a special group has been added to Rewards addon

Reviewer: ```suggestion
To be an administrator of the platform authorized to manage wallets, a special group has been added to Rewards addon
```

Author: Done

which is:

      ::

          /platform/rewarding

> ** Cauri contract  **

Reviewer: ```suggestion
** Token contract  **
```

Author: Done

> The used Cauri contract address can be configured using the following variable:

Reviewer: ```suggestion
The contract address used for the currency token manipulated by wallets can be configured using the following variable:
```

Author: Done

      ::

          exo.wallet.blockchain.token.address=0xc76987d43b77c45d51653b6eb110b9174acce8fb

This address will never change through eXo Platform versions on Ethereum Mainnet blockchain and is shared by all eXo Community.

> The * Cauris * contract is an `ERC-20 <https://eips.ethereum.org/EIPS/eip-20>`__ Token contract that is deployed on Ethereum Mainnet.

Reviewer: ```suggestion
This contract is an `ERC-20 <https://eips.ethereum.org/EIPS/eip-20>`__ Token contract that is deployed on Ethereum Mainnet.
```

Author: Done

> To be able to receive some * Cauris *, a wallet has to be initialized by a * Cauri contract * administrator on blockchain (not on eXo Platform Server bundle that you own).

Reviewer: ```suggestion
To be able to receive some * tokens *, a wallet address must be initialized by a * token contract * administrator on the blockchain (not on the eXo Platform Server bundle that you own).
```

Author: Done

> // TODO ajouter un schéma avec PLUSIEURS exo déployés et au centre la blockchain où on trouve le CONTRAT CAURI (A DEMANDER AU DESIGNER de t'aider !!!!)

Reviewer: ```suggestion
// TODO add a diagram with MANY eXo instances deployed and the blockchain in the center where the token contract lies (ASK DESIGNER HELP)
```

Author: Done
